src/py/pyodide/_package_loader.py

```diff
@@ -16,38 +16,97 @@
 TAR_TYPES = {".tar", ".gz", ".bz", ".gz", ".tgz", ".bz2", ".tbz2"}
 
 
-def unpack_buffer(filename: str, buffer: JsProxy, target: str = "site") -> JsProxy:
+def make_whlfile(*args, owner=None, group=None, **kwargs):
+    return shutil._make_zipfile(*args, **kwargs)  # type: ignore[attr-defined]
+
+
+if IN_BROWSER:
+    shutil.register_archive_format("whl", make_whlfile, description="Wheel file")
+    shutil.register_unpack_format(
+        "whl", [".whl", ".wheel"], shutil._unpack_zipfile, description="Wheel file"  # type: ignore[attr-defined]
+    )
+
+
+def get_format(format: str) -> str:
+    for (fmt, extensions, _) in shutil.get_unpack_formats():
+        if format == fmt:
+            return fmt
+        if format in extensions:
+            return fmt
+        if "." + format in extensions:
+            return fmt
+    raise ValueError(f"Unrecognized format {format}")
+
+
+def unpack_buffer(
+    buffer: JsProxy,
+    *,
+    filename: str = "",
+    format: str = None,
+    target: Literal["site", "lib", None] = None,
+    extract_dir: str = None,
+    calculate_dynlibs: bool = False,
+) -> Optional[JsProxy]:
     """Used to install a package either into sitepackages or into the standard
     library.
 
     This is a helper method called from ``loadPackage``.
 
     Parameters
     ----------
+    buffer
+        A Javascript ``Uint8Array`` with the binary data for the archive.
+
     filename
         The name of the file we are extracting. We only care about it to figure
-        out whether the buffer represents a tar file or a zip file.
+        out whether the buffer represents a tar file or a zip file. Ignored if
+        format argument is present.
 
-    buffer
-        A Javascript ``Uint8Array`` with the binary data for the archive.
+    format
+        Controls the format that we assume the archive has. Overrides the file
+        extension of filename.
+
+    extract_dir
+        Controls which directory the file is unpacked into. Default is the
+        working directory. Mutually exclusive with target.
 
     target
-        Controls which directory the package is installed into. Either "site"
-        which installs the package into the sitepackages directory or "lib"
-        which installs the package into the standard library.
+        Controls which directory the file is unpacked into. Either "site" which
+        unpacked the file into the sitepackages directory or "lib" which
+        unpacked the file into the standard library. Mutually exclusive with
+        extract_dir.
+
+    calculate_dynlibs
+        If true, will return a Javascript Array of paths to dynamic libraries
```

Comment:
Maybe return a list rather, and then call to_js where we need it?

Reply:
I guess we could do that. This is going to be more convenient since otherwise in Javascript we have to do:
```js
const dynlibspy = unpack_archive(...);
const dynlibs = dynlibspy.toJs();
dynlibspy.destroy();
```
which adds two extra lines of code. Also, note that if not `IN_BROWSER`, `to_js` is defined to be the identity function so this doesn't cause trouble for testing.

Reply:
I am going to leave this as-is for now, we could think about changing it later if you have a strong opinion or a good reason arises.

Reply:
Since it's a private function, OK to keep as is.